// appium-logger/build/test/helpers.js
'use strict';

var _core = require('babel-runtime/core-js')['default'];

var _interopRequireWildcard = require('babel-runtime/helpers/interop-require-wildcard')['default'];

_core.Object.defineProperty(exports, '__esModule', {
  value: true
});

var _chai = require('chai');

var _chai2 = _interopRequireWildcard(_chai);

var _sinon = require('sinon');

var _sinon2 = _interopRequireWildcard(_sinon);

var _import = require('lodash');

var _import2 = _interopRequireWildcard(_import);

var _getLogger = require('../lib/logger');

_chai2['default'].should();

function setupWriters() {
  return { stdout: _sinon2['default'].spy(process.stdout, 'write'),
    stderr: _sinon2['default'].spy(process.stderr, 'write') };
}

function getDynamicLogger(testingMode) {
  process.env._TESTING = testingMode ? '1' : '0';
  return _getLogger.getLogger();
}

function restoreWriters(writers) {
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = _core.getIterator(_import2['default'].values(writers)), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var w = _step.value;

      w.restore();
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator['return']) {
        _iterator['return']();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }
}

function assertOutputContains(writers, output) {
  var someoneHadOutput = false;
  var _iteratorNormalCompletion2 = true;
  var _didIteratorError2 = false;
  var _iteratorError2 = undefined;

  try {
    for (var _iterator2 = _core.getIterator(_import2['default'].values(writers)), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      var w = _step2.value;

      if (w.calledWith) {
        someoneHadOutput = w.calledWith(output);
        if (someoneHadOutput) break;
      }
    }
  } catch (err) {
    _didIteratorError2 = true;
    _iteratorError2 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion2 && _iterator2['return']) {
        _iterator2['return']();
      }
    } finally {
      if (_didIteratorError2) {
        throw _iteratorError2;
      }
    }
  }

  if (!someoneHadOutput) {
    throw new Error('Expected someone to have been called with: \'' + output + '\'');
  }
}

function assertOutputDoesntContain(writers, output) {
  var _iteratorNormalCompletion3 = true;
  var _didIteratorError3 = false;
  var _iteratorError3 = undefined;

  try {
    for (var _iterator3 = _core.getIterator(_import2['default'].values(writers)), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
      var w = _step3.value;

      _import2['default'].flatten(w.args).should.not.contain(output);
    }
  } catch (err) {
    _didIteratorError3 = true;
    _iteratorError3 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion3 && _iterator3['return']) {
        _iterator3['return']();
      }
    } finally {
      if (_didIteratorError3) {
        throw _iteratorError3;
      }
    }
  }
}

exports.setupWriters = setupWriters;
exports.restoreWriters = restoreWriters;
exports.assertOutputContains = assertOutputContains;
exports.assertOutputDoesntContain = assertOutputDoesntContain;
exports.getDynamicLogger = getDynamicLogger;
